zcash_client_backend/src/zip321.rs

```diff
@@ -0,0 +1,955 @@
+                _otherwise => {}
+            }
+        }
+
+        return Ok(payment);
+    }
+
+    /// Parser that consumes the leading "zcash:[address]" from
+    /// a zip-0321 URI.
+    pub fn lead_addr<'a, P: consensus::Parameters>(
+        params: &'a P,
+    ) -> impl Fn(&str) -> IResult<&str, Option<IndexedParam>> + 'a {
+        move |input: &str| {
+            let (input, _) = tag("zcash:")(input)?;
+            let (input, addr) = map_res(take_until("?"), |i| Address::parse(params, i))(input)?;
+            let primary_addr_param = addr.map(|a| IndexedParam {
+                param: Param::Addr(a),
+                payment_index: 0,
+            });
+
+            Ok((input, primary_addr_param))
+        }
+    }
+
+    /// The primary parser for <name>=<value> query-string
+    /// parameter pair.
+    pub fn zcashparam<'a, P: consensus::Parameters>(
+        params: &'a P,
+    ) -> impl Fn(&str) -> IResult<&str, IndexedParam> + 'a {
+        move |input| {
+            map_res(
+                separated_pair(indexed_name, char('='), recognize(qchars)),
+                move |r| to_indexed_param(params, r),
+            )(input)
+        }
+    }
+
+    /// Extension for the `alphanumeric0` parser which extends that parser
+    /// by also permitting the characters that are members of the `allowed`
+    /// string.
+    fn alphanum_or(allowed: String) -> impl (Fn(&str) -> IResult<&str, &str>) {
+        move |input| {
+            input.split_at_position_complete(|item| {
+                let c = item.as_char();
+                !(c.is_alphanum() || allowed.contains(c))
+            })
+        }
+    }
+
+    pub fn qchars(input: &str) -> IResult<&str, &str> {
+        alphanum_or("-._~!$'()*+,;:@%".to_string())(input)
+    }
+
+    pub fn namechars(input: &str) -> IResult<&str, &str> {
+        alphanum_or("+-".to_string())(input)
+    }
+
+    pub fn indexed_name<'a>(input: &'a str) -> IResult<&'a str, (&'a str, Option<&'a str>)> {
+        let paramname = recognize(tuple((alpha1, namechars)));
+
+        tuple((
+            paramname,
+            opt(preceded(
+                char('.'),
+                recognize(tuple((one_of("123456789"), digit0))),
+            )),
+        ))(input)
+    }
+
+    pub fn parse_amount<'a>(input: &'a str) -> IResult<&'a str, Amount> {
+        map_res(
+            tuple((digit1, opt(preceded(char('.'), digit1)))),
+            |(whole_s, decimal_s): (&str, Option<&str>)| {
+                let coins: i64 = whole_s
+                    .to_string()
+                    .parse::<i64>()
+                    .map_err(|e| e.to_string())?;
+
+                let zats: i64 = match decimal_s {
+                    Some(d) => format!("{:0<8}", d)
+                        .parse::<i64>()
+                        .map_err(|e| e.to_string())?,
+                    None => 0,
+                };
+
+                let amt = coins * COIN + zats;
```

Comment:
Amount::from_nonnegative_i64, just below here, takes care of  that.

Reply:
How so? I mean that the expression `coins * COIN + zats` can overflow the range of `i64`.

Reply:
The limitation to MAX_MONEY, at least, is taken care of by the from_nonnegative_i64 call. Parse::<i64> fails in the case of values that are too large for an i64. But I do see the point, a weird URI could use the multiplication by COIN to overflow, and even though negative values are excluded by the later parse, I'd have to do the math to see if it could overflow all the way back in to small positive values; maybe it's better to just explicitly limit the coins value here. I'll limit the zats value to at most 8 characters in the parser, so that will eliminate that possibility here.

Reply:
Strictly speaking, `21000000.1` for example is not rejected when it should be.